### PR TITLE
[ARITH] Analyzer RewriteSimplifier: add/sub/mul/div/mod

### PR DESCRIPTION
**Pull request by tqchen**
This PR contains one step of #2588 

- [x] RewriteSimplifier Infra
- [x] Rules for add/sub/mul/div/mod
- [x] Test-cases covering the rules.

We still need to add rules for other ops. Given that these rules are self-contained and covered by test-case, I will request reviews and aim to merge them in before we can in parallel add support for other ops.

Need a few reviewers to check since the rules can be tricky(even though when each of them are covered by at least one test case)

### CODE REVIEW

**Comment by tqchen** [comment]
cc @sgrechanik-h @icemelon9 @merrymercy @kazum @grwlf @ZihengJiang @wweic please help review the PR

**Comment by merrymercy** [comment]
This fails for the case in https://github.com/dmlc/tvm/issues/1711#issuecomment-421063082
```python
ck.verify(x/6*6 + (((x/3) % 2)*3) + (x % 3), x)
```

**Comment by tqchen** [comment]
@merrymercy Rewrite simplifier is hard for solving cases like the nested fusion. We will need a improved canonical simplifier to deal with that. So that specific case will need to wait for a better canonical simplifier

File: src/arithmetic/rewrite_simplify.cc [hunk 1]
```diff
@@ -0,0 +1,646 @@
...
+              const Expr& info,
+              bool override) {
+    if (!override) {
+      CHECK(!var_map_.count(var));
+    }
+    var_map_[var] = info;
+  }
+
+  // Run simplification in post order
+  Expr PostOrderSimplify(Expr expr, int max_iter = 2) {
```
Comment (MarisaKirisame) [inline src/arithmetic/rewrite_simplify.cc]: const Expr& for consistency and efficency

File: src/arithmetic/rewrite_simplify.cc [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+    Expr res = Mutate(x);
+    if (const auto* ptr = res.as<ir::IntImm>()) {
+      return ptr->value == val;
+    }
+    return false;
+  }
+  // Recursive rewrite x
+  // we limit maximum number of recursive rewrite allowed to
+  // avoid infinite loop
+  Expr RecursiveRewrite(Expr x) {
```
Comment (MarisaKirisame) [inline src/arithmetic/rewrite_simplify.cc]: const Expr&

File: src/arithmetic/rewrite_simplify.cc [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+                    broadcast(x * y, lanes));
+    TVM_TRY_REWRITE(ramp(b1, s1, lanes) * broadcast(x, lanes),
+                    ramp(b1 * x, s1 * x, lanes));
+    TVM_TRY_REWRITE(broadcast(x, lanes) * ramp(b1, s1, lanes),
+                    ramp(b1 * x, s1 * x, lanes));
+  }
+
+  if (IsIndexType(op->type)) {
+    // constant simplification rule
+    TVM_TRY_REWRITE((x + c1) * c2, x * c2 + c1 * c2);
```
Comment (grwlf) [inline src/arithmetic/rewrite_simplify.cc]: Do this rule contradicts with `TVM_TRY_REWRITE(x * y + x * z, x * (y + z))` from the `Add` section? Does RewriteSimplifier stabilizes if repeatedly applied many times?

File: tests/python/unittest/test_arith_rewrite_simplify.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
+import tvm
```
Comment (grwlf) [inline tests/python/unittest/test_arith_rewrite_simplify.py]: I think that this kind of tests is not enough for such a complex algorithm. It is hard to be sure that existing test cases trigger every rule from the long list above. Probably we have to generate test cases with answers rather then hardcode them.

File: src/arithmetic/rewrite_simplify.cc [hunk 5]
```diff
@@ -0,0 +1,646 @@
...
+#include <tvm/expr_operator.h>
+#include <tvm/ir_mutator.h>
+#include "const_fold.h"
+#include "pattern_match.h"
+
+namespace tvm {
+namespace arith {
+
+using namespace ir;
+
```
Comment (grwlf) [inline src/arithmetic/rewrite_simplify.cc]: In my opinion, this approach to specifying rewriting rules has pros and cons. The good thing is that rule definition looks really clear. The bad thing I see is that rules are declared one after another in a single C++ function. We can't combine rules into bigger rules, and, more importantly, we can't call them individually, for example in order to apply a testcase.

File: src/arithmetic/rewrite_simplify.cc (hunk 3 above)
Comment (tqchen) [reply src/arithmetic/rewrite_simplify.cc]: This rewrite only works for constant and note that c1 * c2 will get automatically constant folded

**Comment by tqchen** [comment]
Thanks, @kazum @merrymercy @grwlf @sgrechanik-h for reviews, to collectivly answer some of the questions

## Coverage of Testcases

We do not expect RewriteSimplifier to simplify all the possible cases. As a matter of fact, because most rewriting are **local** it is hard to get simplification over a long distance. This is a known limitation of the rewrite-based simplification. 

To complement that, we will need a new simplification strategy: CanonicalSimplifier, which brings canonical form and will make it much easier to simplify cases like ```(x + y + z + x)=> x*2 + y + z```. However, there are cases that are covered by RewriteSimplifier which may not be covered by CanonicalSimplifiers, in particular, the rules wrt to max/mins.

So the goal of this PR is to get something that is correct and off the ground. We will continue to bring in new CanonicalSimplifier and improve both collectively.

## Correctness

Correctness is a big problem for us, that is why we need test-coverage, and code reviews to review these rules. The general goal is to vet the rules explicitly. Again since many of the weights are going to be carried by CanonicalSimplifier, we do not need to have a comprehensive list of rules in here.

File: src/arithmetic/rewrite_simplify.cc (hunk 5 above)
Comment (tqchen) [reply src/arithmetic/rewrite_simplify.cc]: in theory, CSE(common subexpression elimination) in compiler could pull out some of the common computed element of these rules and not recompute them

File: src/arithmetic/rewrite_simplify.cc [hunk 6]
```diff
@@ -0,0 +1,646 @@
...
+    TVM_TRY_RECURSIVE_REWRITE(x - y * c1, x + y * (0 - c1));
+  }
+
+  // condition rules.
+  TVM_TRY_REWRITE(select(x, b1, b2) - select(x, s1, s2),
+                  select(x, b1 - s1, b2 - s2));
+  TVM_TRY_REWRITE(select(x, y, z) - z,
+                  select(x, y - z, ZeroWithTypeLike(z)));
+  TVM_TRY_REWRITE(select(x, y, z) - y,
+                  select(x, ZeroWithTypeLike(y), z - y));
```
Comment (kazum) [inline src/arithmetic/rewrite_simplify.cc]: Looks like tests for these cases are missing.
```
ck.verify(tvm.expr.Select(x > 0, y, z) - y, tvm.expr.Select(x > 0, 0, z - y))
ck.verify(tvm.expr.Select(x > 0, y, z) - z, tvm.expr.Select(x > 0, y - z, 0))
```

File: src/arithmetic/rewrite_simplify.cc (hunk 1 above)
Comment (tqchen) [reply src/arithmetic/rewrite_simplify.cc]: Need to mutate expr during the process, so keep it non-const, will change the RecursiveRewrite one

**Comment by tqchen** [comment]
Changed to limit recursive depth as per suggestion by @wweic and @sgrechanik-h , added test cases suggested by @kazum . Please help take another look

**Comment by tqchen** [comment]
Thanks @kazum @wweic @sgrechanik-h @grwlf , this is now merged

**Comment by tqchen** [comment]
Follow up PR

**Pull request merged**